services/pull/patch.go

```diff
@@ -184,11 +186,88 @@ func TestPatch(pr *models.PullRequest) error {
 		if conflict {
 			pr.Status = models.PullRequestStatusConflict
 			log.Trace("Found %d files conflicted: %v", len(pr.ConflictedFiles), pr.ConflictedFiles)
+
 			return nil
 		}
 		return fmt.Errorf("git apply --check: %v", err)
 	}
+
+	if pr.Index != 0 {
+		if err = CheckPullFilesProtection(pr); err != nil {
+			return fmt.Errorf("pr.CheckPullFilesProtection(): %v", err)
+		}
+	}
+
+	if len(pr.ChangedProtectedFiles) > 0 {
+		log.Trace("Found %d protected files changed", len(pr.ChangedProtectedFiles))
+	}
+
 	pr.Status = models.PullRequestStatusMergeable
 
 	return nil
 }
+
+// CheckFileProtection check file Protection
+func CheckFileProtection(oldCommitID, newCommitID string, patterns []glob.Glob, limit int, repo *git.Repository) ([]string, error) {
+	stdout, err := git.NewCommand("diff", "--name-only", oldCommitID+"..."+newCommitID).RunInDir(repo.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(patterns) == 0 {
+		return nil, nil
+	}
+
+	var changedFiles []string
+	if limit <= 10 {
+		changedFiles = make([]string, 0, limit)
+	} else {
+		changedFiles = make([]string, 0, 10)
+	}
+
+	for _, path := range strings.Split(stdout, "\n") {
+		lpath := strings.ToLower(strings.TrimSpace(path))
+		for _, pat := range patterns {
+			if pat.Match(lpath) {
+				changedFiles = append(changedFiles, path)
+				break
+			}
+		}
+
+		if len(changedFiles) >= limit {
+			break
+		}
+	}
+
+	return changedFiles, nil
+}
```

Comment:
Please use similar code to that in `cmd/hook.go:checkFileProtection(...)`.

The number of files changed in a diff can be arbitrarily large - storing the stdout to just iterate over them and throw most of them away is just wasteful.

Reply:
Hmm... looking on this needs more things changed here.

* We should be using the temporary repository context in line 89 here and the diff that generates the patch from that to check for protected file changes.
*